kasa/protocol.py

```diff
@@ -28,6 +29,46 @@
 _NO_RETRY_ERRORS = {errno.EHOSTDOWN, errno.EHOSTUNREACH, errno.ECONNREFUSED}
 _UNSIGNED_INT_NETWORK_ORDER = struct.Struct(">I")
 
+_T = TypeVar("_T")
+
+
+def redact_data(data: _T, redactors: dict[str, Callable[[Any], Any] | None]) -> _T:
+    """Redact sensitive data for logging."""
+    if not isinstance(data, (dict, list)):
+        return data
+
+    if isinstance(data, list):
+        return cast(_T, [redact_data(val, redactors) for val in data])
+
+    redacted = {**data}
+
+    for key, value in redacted.items():
+        if value is None:
+            continue
+        if isinstance(value, str) and not value:
+            continue
+        if key in redactors:
+            if redactor := redactors[key]:
+                try:
+                    redacted[key] = redactor(value)
+                except:  # noqa: E722
+                    redacted[key] = "**REDACTEX**"
```

Comment:
```suggestion
                    redacted[key] = "**REDACTED**"
```

Reply:
This is `REDACTEX` so we know if we see it in the logs that the redaction method threw an exception.

Reply:
Perhaps debug log it, too? It shouldn't be hit that often, so having the information available saves a round-trip if it ever happens.

Reply:
I did think about this but I wanted to keep the redactor as resilient and simple as possible. It should be fairly obvious what type of error we would have gotten from the lambda itself and if we log more than the just the type we may end up having to redact the redactor 😃